### DISCUSSION TITLE
Access facial recognition from api

### DISCUSSION THREAD

**Discussion by jsapede** [immich-app/immich#12916] [Feature Request]
Would it be possible to run facial récognition from an image sent through the api ?
The Idea would be to have coupling with home assistant and frigate nvr.
As soon as frigate détect a person, it sends mqtt advertising to homeassistant including Snapshot. Forwarding this Snapshot to immich in a specific asset , for exemple named "frigate" and Ask immich to run facial récognition on this Snapshot and get back the identity from the api response.

**Comment by bo0tzz** [comment]
#9285

**Comment by arne182** [comment]
But this is not the same as 9285. Maybe we should start by uploading through home assistant first. Let me see if I can get that working

**Reply by bo0tzz** [reply]
It is the same in essence, 9285 would include the API that is requested here.

**Reply by jsapede** [reply]
> But this is not the same as 9285. Maybe we should start by uploading through home assistant first. Let me see if I can get that working

i've made some tries but my coding / templating/python levels are very weak.

however the proof of concept works : 

1/ upload an asset using curl or python exemple 
2/ retrieve the id of the asset in the curl/python response 
3/ use api getfaces with the retrieved api id and parse the json response to get name.

i built a light script to retrieve this in homeassistant using restful command with GET

```
sequence:
  - action: rest_command.immich_reco
    data: {}
    response_variable: immich_response
  - variables:
      face: "{{ immich_response['content'][0]['person']['name']}}"
alias: immich-request-face
description: ""
```

made some tests with some pictures and it works perfectly

the only need for "surveillance" with frigate is to get sublabel on known faces, so immich has to be trained with asset collections on the face you want to recognise. there's no need to have specific api as soon as getfaces gives the answer.

maybe my question was wrong as getfaces seems to supply already the informations we want and facial récognition is automatically activated in the settings for the new assets.

i'm looking forward to automate this using restful commands but uploading pictures with rest doesn't seem to work.

Maybe using python exemples from the immich API doc as scripts to automate the.process on camera détection should be the easier way. Will try to build something.

**Reply by PriestlyHall** [reply]
I have this working thanks to op's framework. The problem I have now is latency between uploading a new image to Immich, and Immich processing the new asset for facial recognition. If I query getfaces during this latency, it incorrectly responds with 0 matches. I'm putting a 3s delay to deal with this, but is there a way to check if an asset has been processed, or for immich to send a rest command to home assistant notifying of the same?

**Reply by PriestlyHall** [reply]
Currently looking at getAllJobsStatus api call, seems to be what I need to confirm face detection has been completed. Need to poll it, confirm the count for face detection is zero, then run getfaces.

**Comment by ferbar** [comment]
Hi, has anybody tried to send a photo directly to the immich machine learning api?
the machine learning should have a `/predict` api call:
https://github.com/immich-app/immich/blob/main/machine-learning/immich_ml/main.py#L165

`output = await run(model.predict, *inputs, **entry["options"])`
https://github.com/immich-app/immich/blob/main/machine-learning/immich_ml/main.py#L195

I didn't find any usefull documentation what are the input parameters and what is `output`
`inputs` can at least be an image. Since everything in immich is user based, I think there should be an user-id, too. ... but maybe i'm completly wrong ...

**Reply by cmprmsd** [reply]
@ferbar I had the same issue some years ago (https://github.com/immich-app/immich/discussions/14340) and solved it with an additional service that you add to the stack. This provides access to the predict endpoint and the necessary database access to match the embeddings.

https://github.com/cmprmsd/immich-face-api

**Comment by mcattle** [comment]
I, too, would like to access facial recognition through the official Immich API.  I'd like to be able to submit a picture to the API (without uploading it to Immich itself for storage), and get back a list of known faces in the image, perhaps with confidence percentages (or set the minimum required confidence level in the request).  This way, I can leverage the trained database of people I know to detect who is at the door in Home Assistant.

I suggest reopening this feature request for this use case, as it would then allow the Home Assistant Immich integration to officially support it out of the box.

In the meanwhile, I'll look into the additional service linked by @cmprmsd as a workaround, albeit lacking the security of an Immich API Key.
